### PR TITLE
Minimize Prop <= i to i := Set, restrict minimization to set to flexibles

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix part of #8196
Replaces #9343

### CODE REVIEW

**Comment by ppedrot** [comment]
Literally no one on earth can review this PR except from the two guys who wrote it at the CUDW. @maximedenes will have the pleasure to be assigned to this PR, I am not match.

**Comment by SkySkimmer** [comment]
@mattam82 equations is broken, only you can fix it

**Comment by mattam82** [comment]
Indeed, looks like a legitimate mishandling of universes in Equations

**Comment by maximedenes** [comment]
@mattam82 any chance you can have a look at the Equations failure?

**Comment by maximedenes** [comment]
@SkySkimmer I'll rebase and push

**Comment by maximedenes** [comment]
@Janno Thanks to @mattam82, we now have a fix for Equations. Would you mind fixing Mtac2 w.r.t. this patch? Then we can merge it!

File: doc/changelog/02-specification-language/10331-minim-prop-toset.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- Changed heuristics for universe minimization to Set: only minimize
```
Comment (Zimmi48) [inline doc/changelog/02-specification-language/10331-minim-prop-toset.rst]: Should be ``` :g:`Set` ``` or ``` :math:`\Set` ```

File: doc/changelog/02-specification-language/10331-minim-prop-toset.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+  flexible universes, and also use constraints ``Prop <= i`` `#10331
```
Comment (Zimmi48) [inline doc/changelog/02-specification-language/10331-minim-prop-toset.rst]: Add parentheses around the meta-information on PR and authorship.

**Comment by SkySkimmer** [comment]
What's the equations fix?

**Comment by maximedenes** [comment]
> What's the equations fix?

https://github.com/mattam82/Coq-Equations/pull/219

**Comment by Janno** [comment]
Okay, it took me quite a while but I have a hypothesis as to what's going on in tests/dummylang.v of Mtac2: some definitions within the file have universes that cannot be unified with `Set`. Without going into too much detail, it seems that monomorphic definitions always end up with all their universes bigger than `Set` whereas polymorphic ones can have universes equal to `Set`. Maybe I should know this and also why this is the case but I am a bit puzzled by this right now.

So this bug can be fixed by making some of the definitions polymorphic. Weird, but OK. I would prefer a more canonical solution but we could fix it this way, for now.

tests/bugs.v is a different issue. Type class search fails to find an instance for a type class (`seq_one` for class `Seq`). What's weird here is that the issue can be solved either by introducing this Hint
```coq
Hint Extern 0 (Seq _ _ _) => apply seq_one : typeclass_instances.
```
or by disabling `Minimization ToSet` around the Instance declaration. I checked the difference in universes and it seems that there is exactly one universe that is correctly being minimized to `Set` because the instance instantiates the corresponding type with a `Prop`. However, I do not understand why the Hint fixes this issue. If the hint works, why could type class search fail?

For that bug I don't find either fix acceptable. How can I debug this further? I don't understand the output of `Set Debug Unification` and neither do I know if that actually prints unification triggered by type class search.

**Comment by SkySkimmer** [comment]
>Without going into too much detail, it seems that monomorphic definitions always end up with all their universes bigger than Set whereas polymorphic ones can have universes equal to Set.

Making a monomorphic universe `= Set` is a strong effect: it prevents having an universe below it. Instantiating a polymorphic universe to Set is local to the specific instantiation.

In other words:
~~~coq
Definition foo := Type.
Definition bar := fun x:foo => x:Set.
Definition baz := Type : foo. (* one of these commands must fail *)
~~~
Having `baz` fail is considered surprising / undesirable so we prevent it by forcing foo > Set.
If foo is polymorphic then there's no problem so we don't force it > Set.

**Comment by Janno** [comment]
> Having baz fail is considered surprising / undesirable so we prevent it by forcing foo > Set.
If foo is polymorphic then there's no problem so we don't force it > Set.

That makes a lot of sense. Thank you for explaining it!

The only remaining issue is the failure in tests/bug.v. I still don't know what to do about it.

**Comment by SkySkimmer** [comment]
CI only shows a failure in tests/dummylang.v, is that the one you mean?

**Comment by Janno** [comment]
No, I think the CI just doesn't run `make` with `-k` so it stops at the first error, which happens to be dummylang.v.

I seem to have lost my opam switch with this branch in it so I can't test it easily now but IIRC the other failure was at the very end of tests/bugs.v, where the `&>` type class instance can't be resolved.

**Comment by SkySkimmer** [comment]
Let's try splitting the 2 changes of this PR apart, maybe it'll be clearer what's going on.

**Comment by Janno** [comment]
This reduced PR has both Mtac2 failures from the combined one *plus* an additional universe inconsistency in tests/sf-5/lf/Tactics.v involving `Set`. I suppose an inflexible universe was minimized to `Set`, which would make sense given that the restriction to flexible universes was moved to #10657.

**Comment by Janno** [comment]
I just saw the 8.11 schedule. Is there any chance of getting this PR into 8.11? I am willing to change the Mtac2 code to work around the two issues if that would make this PR eligible for inclusion in 8.11.

(I don't like adding these workarounds because I don't understand them but I'd rather have reasonably fast Mtac2 with workarounds than not having fast Mtac2 at all.)

**Comment by SkySkimmer** [comment]
I don't really have time to focus on this, anyone who can push it forward please feel free (@mattam82 maybe?)

**Comment by mattam82** [comment]
Looks like elpi just needs an API change, equations requires some more new universe annotations and the test-suite just fails. I can have a quick look tomorrow (don't hesitate to ping me if I don't).

**Comment by maximedenes** [comment]
@SkySkimmer I guess we should merge https://github.com/coq/coq/pull/10657, then rebase this, right?

**Comment by SkySkimmer** [comment]
>@SkySkimmer I guess we should merge #10657, then rebase this, right?

Sounds reasonable.
It's been a while since I looked at these 2 PRs so I don't remember the status.

**Comment by maximedenes** [comment]
@SkySkimmer #10657 merged, would you mind rebasing?

**Comment by SkySkimmer** [comment]
Rebased

File: doc/changelog/02-specification-language/10331-minim-prop-toset.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+- Changed heuristics for universe minimization to :g:`Set`: also use
```
Comment (Zimmi48) [inline doc/changelog/02-specification-language/10331-minim-prop-toset.rst]:
```suggestion
- **Changed:** Heuristics for universe minimization to :g:`Set`: also use
```

**Review by Zimmi48 — COMMENTED** [review]
I suppose this was not documented in enough details to require an update to the relevant chapters of the documentation?

File: doc/changelog/02-specification-language/10331-minim-prop-toset.rst (hunk 3 above)
Comment (mattam82) [reply doc/changelog/02-specification-language/10331-minim-prop-toset.rst]: It actually is closed to what the refman says: only "fresh" universes are minimized

**Comment by maximedenes** [comment]
Wow:

```
File "./theories/Logic/Hurkens.v", line 437, characters 0-4:
Error: Anomaly "File "kernel/univ.ml", line 875, characters 4-10: Assertion failed."
```

**Comment by maximedenes** [comment]
@SkySkimmer any idea what this anomaly is about?

File: engine/univMinim.ml [hunk 4]
```diff
@@ -273,12 +273,13 @@ let normalize_context_set ~lbound g ctx us algs weak =
   let (ctx, csts) = ContextSet.levels ctx, ContextSet.constraints ctx in
   (* Keep the Prop/Set <= i constraints separate for minimization *)
   let smallles, csts =
-    Constraint.partition (fun (l,d,r) -> d == Le && (Level.equal l lbound || Level.is_sprop l)) csts
+    Constraint.partition (fun (l,d,r) -> d == Le && (Level.equal l lbound || Level.is_prop l || Level.is_sprop l)) csts
```
Comment (SkySkimmer) [inline engine/univMinim.ml]: The lbound use here was incorrect.

Comment (SkySkimmer) [reply engine/univMinim.ml]: @mattam82 maybe we shouldn't use lbound and instead just separate based on is_small (as it used to be)?

Comment (mattam82) [reply engine/univMinim.ml]: That sounds right indeed.

**Comment by maximedenes** [comment]
@SkySkimmer this breaks the test-suite, it seems. It wasn't expected, was it?

**Comment by SkySkimmer** [comment]
Looks like a bad interaction with the lbound stuff, for instance 
~~~coq
(** success/Inductive.v *)
Inductive L (A:Type) (T:=A) : Type := C : L nat -> L A.
(* 
Error: Illegal application: 
The term "L" of type "forall A : Type, let T := A in Prop"
cannot be applied to the term
 "nat" : "Set"
This term has type "Set" which should be coercible to 
"Type".
*)

(** bug_5683.v *)

Inductive term A :=
| term_abs : ty -> term (Succ A) -> term A
| term_app : term A -> term A -> term A
| term_var : A -> term A
| term_nat : nat -> term A.
(* 
File "./bugs/closed/bug_5683.v", line 16, characters 0-155:
Error: Illegal application: 
The term "term" of type "Type -> Type"
cannot be applied to the term
 "Succ A" : "Type"
This term has type "Type@{max(Set,bug_5683.13)}" which should be coercible to
 "Type@{bug_5683.13}".
*)
~~~
in both cases it fails to coerce Set to Type. I guess my first fix was the correct one, we can't use is_small when lbound is Prop because it may forget Set<=i constraints.

**Comment by SkySkimmer** [comment]
Now we get a whole new error x_x

**Comment by maximedenes** [comment]
@SkySkimmer do you think we need to discuss this with @mattam82? I'm afraid I can't help much technically, I don't understand this code.

**Comment by SkySkimmer** [comment]
After much debug: the current error goes away if we make CRealLt_morph univ poly.
This is likely an issue with using `push_context_set ~strict:true`

**Comment by SkySkimmer** [comment]
Other notes: unification.ml doesn't produce reasons for unification errors, and typeclass debug doesn't print reasons because the reason printer is higher up in himsg

**Comment by Janno** [comment]
Is there still hope for this PR? :)

I think the additional universes generated for Props are still the number one reason for subpar Mtac2 performance.

**Comment by SkySkimmer** [comment]
TBH I don't remember what was going on with this PR, if we still want it we may as well start it over so I'm closing.

**Pull request closed without merging**